Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,671 @@
+    /// </summary>
+    public LineVerificationStatus[] PerLineDiagnostic { get; init; }
+
+    static LineVerificationStatus[] RenderPerLineDiagnostics(
+      VerificationDiagnosticsParams verificationDiagnosticsParams,
+      NodeDiagnostic[] perNodeDiagnostic,
+      int numberOfLines,
+      int numberOfResolutionErrors,
+      bool verificationStarted,
+      Container<Diagnostic> diagnostics) {
+      var result = new LineVerificationStatus[numberOfLines];
+
+      if (perNodeDiagnostic.Length == 0 && numberOfResolutionErrors == 0 && verificationStarted) {
+        for (var line = 0; line < numberOfLines; line++) {
+          result[line] = LineVerificationStatus.Verified;
+        }
+
+        return result;
+      }
+
+      // Render node content into lines.
+      foreach (var nodeDiagnostic in perNodeDiagnostic) {
+        if (nodeDiagnostic.Filename == verificationDiagnosticsParams.Uri.GetFileSystemPath() ||
+            "untitled:" + nodeDiagnostic.Filename == verificationDiagnosticsParams.Uri) {
+          nodeDiagnostic.RenderInto(result);
+        }
+      }
+
+      // Fill in the missing "Unknown" based on the surrounding content
+      // The filling only takes Verified an Error
+      var previousNotUnknown = LineVerificationStatus.Unknown;
+      var lineDelta = 1;
+      // Two passes so that we can fill gaps based on what happened before AND after
+      for (var line = 0; 0 <= line; line += lineDelta) {
+        if (line == numberOfLines) {
+          lineDelta = -1;
+          previousNotUnknown = LineVerificationStatus.Unknown;
+          continue;
+        }
+        if (previousNotUnknown != LineVerificationStatus.Verified &&
+            previousNotUnknown != LineVerificationStatus.VerifiedObsolete &&
+            previousNotUnknown != LineVerificationStatus.VerifiedVerifying) {
+          previousNotUnknown = LineVerificationStatus.Unknown;
+        }
+        if (result[line] == LineVerificationStatus.Unknown) {
+          result[line] = previousNotUnknown;
+        } else {
+          previousNotUnknown = result[line];
+        }
+      }
+
+      var resolutionErrorRendered = 0;
+      foreach (var diagnostic in diagnostics) {
+        if (resolutionErrorRendered >= numberOfResolutionErrors) {
+          break;
+        }
+        result[diagnostic.Range.Start.Line] = LineVerificationStatus.ResolutionError;
+        resolutionErrorRendered++;
+      }
+
+      var existsErrorRange = false;
+      var existsError = false;
+      foreach (var line in result) {
+        existsErrorRange = existsErrorRange || line == LineVerificationStatus.ErrorRange;
+        existsError = existsError || line == LineVerificationStatus.Error;
+      }
+
+      if (existsErrorRange && !existsError) {
+        existsError = false;
+      }
+
+      return result;
+    }
+  }
+
+
+  public enum VerificationStatus {
+    Unknown = 0,
+    Verified = 200,
+    Inconclusive = 270,
+    Error = 400
```

Comment:
Is there any reasoning behind the specific numbers chosen here?

Reply:
Yes although these numbers are totally arbitrary, there are several causes to them.
1) HTTP codes 200 means OK and HTTP code 400 means error.
2) I wanted that the linear status code be roughly the sum of this verification code and the status code.
3) One of these two codes has to be a higher number than the other, and it made sense that verified + pending is lower than error + current, although pending is greater (more up-to-date) than current.
3) Code 300 is currently used to report various error ranges (see LinearVerificationStatus)
4) I am not using Inconclusive right now, and I wanted something between verified and error range, but closer to error range.

Reply:
Seems reasonable!